snuba/datasets/plans/single_table.py

```diff
@@ -0,0 +1,70 @@
+import itertools
+from typing import Sequence
+
+from snuba.datasets.plans.query_plan import (
+    QueryPlanExecutionStrategy,
+    SingleQueryRunner,
+    StorageQueryPlan,
+    StorageQueryPlanBuilder,
+)
+from snuba.datasets.storage import QueryStorageSelector, Storage
+from snuba.query import RawQueryResult
+from snuba.query.query_processor import QueryProcessor
+from snuba.request import Request
+
+
+class SimpleQueryPlanExecutionStrategy(QueryPlanExecutionStrategy):
+    def execute(self, request: Request, runner: SingleQueryRunner) -> RawQueryResult:
+        return runner(request)
+
+
+class SingleTableQueryPlanBuilder(StorageQueryPlanBuilder):
+    """
+    Builds the Storage Query Execution Plan for a dataset that is based on
+    a single storage.
+    """
+
+    def __init__(
+        self, storage: Storage, post_processors: Sequence[QueryProcessor]
+    ) -> None:
+        # The storage the query is based on
+        self.__storage = storage
+        # This is a set of query processors that have to be executed on the
+        # query after the storage selection but that are defined by the dataset.
+        # Query processors defined by a Storage must be executable independently
+        # from the context the Storage is used (whether the storage is used by
+        # itself or whether it is joined with another storage).
+        # In a joined query we would have processors defined by multiple storages.
+        # that would have to be executed only once (liek Prewhere). That is a
+        # candidate to be added here as post process.
+        self.__post_processors = post_processors
```

Comment:
The key concrete issue I had without this is that, we would end up running some query processors twice (like Prewhere) if we were not able to apply some query processors after the storage ones and only once per query. We could make all of them idempotent but that would be quite hard to enforce.